### DISCUSSION TITLE
How to reconnect to the opc server when we lost connection with the opc server ?

### DISCUSSION THREAD

**Discussion by argiraud** [eclipse-milo/milo#833] [General]
Hi,
I'm trying to test if my app reconnect to the opc server when I plug out my app from the network and plug in after 10 seconde. It appears that subscriptions can't be re-create unless I re-run my app. I'm in version 0.6.1.
I saw that normally the re-connection is automatic.
I don't know what I do wrong.

![Sans titre](https://user-images.githubusercontent.com/33487226/116562101-2d017b80-a903-11eb-9f33-23a40bdd4362.png)

**Comment by kevinherron** [comment]
All you need to do is add a `SubscriptionListener` to the `OpcUaSubscriptionManager` and implement `SubscriptionListener::onSubscriptionTransferFailed`. Any time this is called it means the client has reconnected but was unable to resume the session or transfer the old subscription, so you have to re-create it.

**Reply by argiraud** [reply]
Thanks for your reactivity! You were really helpful and it resolve my problem.

**Comment by argiraud** [comment]
I'm having issues when my opc-client trying to resubscribe to the Matrikon OPC server. When I lose the connection and try to reconnect subscription with onSubscriptionTransferFailed the Matrikon OPC server is going too busy so it refused to resubscribe. Do I have to wait a little amount of time each time I am trying to recreate subscription in my code ?

**Reply by kevinherron** [reply]
That's not something I've seen before. What kind of error do you get when you create your subscriptions again?

**Reply by argiraud** [reply]
> 10:59:39.129 [milo-shared-thread-pool-12] ERROR c.a.o.l.SessionListenerHandlingSubscription.onSessionInactive(32) - Session inactive for opc.tcp://XXX/MatrikonOpcUaWrapper
> 
> 10:59:40.670 [milo-shared-thread-pool-13] ERROR c.a.o.listener.SubscriptionListener.onSubscriptionTransferFailed(48) - onSubscriptionTransferFailed for opc.tcp://XXX/MatrikonOpcUaWrapper subscription 35 status: StatusCode{n
> ame=Bad_ServiceUnsupported, value=0x800B0000, quality=bad} try to force resubscription
> 
> 10:59:41.200 [milo-shared-thread-pool-12] INFO  c.a.o.l.SessionListenerHandlingSubscription.onSessionActive(21) - Session active for opc.tcp://XXX/MatrikonOpcUaWrapper
> 
> 10:59:41.805 [milo-shared-thread-pool-12] ERROR c.a.o.l.SessionListenerHandlingSubscription.onSessionInactive(32) - Session inactive for opc.tcp://XXX/MatrikonOpcUaWrapper
>

Again and again 
Sometimes it go back to normal, sometimes not and after a certain moment it stops trying to reconnect.
When my opc-client is doing that I can't connect to OPC server with an another opc client like UAExpert and when I stop my opc client UAExpert can connect to the server.

In Matrikon server there are serverTooBusy errors.

The only solution I found is to restart my opc client but it doesn't work all the time.

**Comment by kevinherron** [comment]
What does your re-subscribe logic/code look like? I guess maybe Matrikon
can't handle very much and you might need to build a little delay into the
logic. Are there a lot of subscriptions or monitored items you're making?

**Reply by argiraud** [reply]
```
public void forceResuscribeOnReconnection() throws Exception {
        if (!tagSubscribes.isEmpty()) {
            opcUaClient.getSubscriptionManager().clearSubscriptions();
            List<TagConfiguration> oldSubscriptions = new ArrayList<>(tagSubscribes);
            tagSubscribes.clear();
            subscribe(oldSubscriptions, this.valueConsumer);
        }
    }
public void subscribe(List<TagConfiguration> tags, BiConsumer<TagConfiguration, TagValue> valueConsumer) throws ExecutionException, InterruptedException {
        this.valueConsumer = valueConsumer;
        UaSubscription subscription = opcUaClient.getSubscriptionManager().createSubscription(this.requestedPublishInterval).get();
        List<MonitoredItemCreateRequest> requests = createMonitoredItemCreateRequest(tags);
        createMonitoredItems(tags, subscription, requests);
    }
```

Maybe you right i should build a little delay. There are like one hundred monitored items.

**Reply by kevinherron** [reply]
You don't need that call to `clearSubscriptions`. If transfer fails the subscription manager cleans up its own bookkeeping. 

You will get one callback per failed subscription transfer, so only re-create that particular subscription when it happens, don't just clear everything and re-create everything.

**Reply by argiraud** [reply]
Yes you are right.
I have only one subscription for all my items. So I have to recreate only the monitored items not working ?

**Comment by kevinherron** [comment]
You'll have to re-create the subscription and all its monitored items.

I'm just looking at your logic and if you ever had multiple subscriptions
you would need to make the changes I suggested. If multiple subscriptions
are in use, and the server doesn't support transfers, then you'll get a
callback for each of them. If in each callback you tried to clear and
re-create *all* subscriptions you should be able to see why that would lead
to issues.

**Reply by argiraud** [reply]
I will make the changes and look in my logic to see why I send too many things to Matrikon.
Thanks again for your help !

**Comment by kevinherron** [comment]
I’m going to be away from internet and cell service for a few days. If you get stuck then getting a Wireshark capture of this reconnect dynamic with security disabled might help me see what’s going on better when I’m back.

**Reply by argiraud** [reply]
Thanks I will do it
